### ISSUE TITLE
"free(): invalid pointer" when using CPR with conan

### ISSUE THREAD

**Issue by TruncatedDinosour** [libcpr/cpr#634]
## Description

When using CPR with conan I get this error when trying to run the binary
```
free(): invalid pointer
Aborted
```

## Expected Behavior

Should work

## Actual Behavior

It breaks

## Steps to Reproduce

1. Use CPR with conan
2.
3.
4.

## Context

I was trying to reach out to a website and get the response body/text.

## Your Environment

* Version used: 1.6.2
* Where did you get it from (e.g. conan, vcpkg, master, ...): conan
* Operating System and version: Linux ari-gentoo 5.14.2-gentoo #1 SMP Thu Sep 9 18:26:44 EEST 2021 x86_64 Intel(R) Core(TM) i3-8130U CPU @ 2.20GHz GenuineIntel GNU/Linux
* Link to a small example: https://replit.com/@B00bleaTea/ghufhuerhyur4gyfghryfhryhfyufhurehufberfhberhbfhuerbfhuerf#main.cpp

**Run this when trying the example (in the shell tab):**
```bash
chmod a+rx ./build.sh

./build.sh

./bin/conantest 'https://google.com'
```

**Comment by TruncatedDinosour** [comment]
I closed it because I thought it might have been my issue, but doesn't look like it

**Comment by COM8** [comment]
I can confirm this one. Thanks for reporting.
Looks like this happens just with the conan package.

This happens when calling the default destructor of `cpr::Response`.

**Comment by COM8** [comment]
Since you are using CMake try this instead:
```cmake
cmake_minimum_required(VERSION 3.16)

project(conantest CXX)

add_executable(${PROJECT_NAME} main.cpp)

include(FetchContent)
FetchContent_Declare(cpr GIT_REPOSITORY https://github.com/whoshuu/cpr.git GIT_TAG f4622efcb59d84071ae11404ae61bd821c1c344b) # the commit hash for 1.6.2
set(CPR_BUILD_TESTS OFF CACHE INTERNAL "" FORCE)
set(CURL_CA_FALLBACK ON CACHE INTERNAL "" FORCE) # Workaround for failing SSL requests
FetchContent_GetProperties(cpr)
if(NOT cpr_POPULATED)
  FetchContent_Populate(cpr)
  # Do not add install targets for cpr
  add_subdirectory(${cpr_SOURCE_DIR} ${cpr_BINARY_DIR} EXCLUDE_FROM_ALL)
endif()

target_link_libraries(${PROJECT_NAME} cpr::cpr)
```

```bash
mkdir build
cd build
cmake ..
cmake --build .
```

**Comment by TruncatedDinosour** [comment]
> Since you are using CMake try this instead:
> 
> ```cmake
> cmake_minimum_required(VERSION 3.16)
> 
> project(conantest CXX)
> 
> add_executable(${PROJECT_NAME} main.cpp)
> 
> include(FetchContent)
> FetchContent_Declare(cpr GIT_REPOSITORY https://github.com/whoshuu/cpr.git GIT_TAG f4622efcb59d84071ae11404ae61bd821c1c344b) # the commit hash for 1.6.2
> set(CPR_BUILD_TESTS OFF CACHE INTERNAL "" FORCE)
> set(CURL_CA_FALLBACK ON CACHE INTERNAL "" FORCE) # Workaround for failing SSL requests
> FetchContent_GetProperties(cpr)
> if(NOT cpr_POPULATED)
>   FetchContent_Populate(cpr)
>   # Do not add install targets for cpr
>   add_subdirectory(${cpr_SOURCE_DIR} ${cpr_BINARY_DIR} EXCLUDE_FROM_ALL)
> endif()
> 
> target_link_libraries(${PROJECT_NAME} cpr::cpr)
> ```
> 
> ```shell
> mkdir build
> cd build
> cmake ..
> cmake --build .
> ```

thank you, works

**Comment by KingKili** [comment]
I am reopening this issues since this seems to be a workaround, and no solution.

**Comment by KingKili** [comment]
> I can confirm this one. Thanks for reporting.
> Looks like this happens just with the conan package.
> 
> This happens when calling the default destructor of `cpr::Response`.

I dug a little deeper using a debug build of cpr. To be exact it is happening when the destructor of 'cpr::StringHolder<cpr::Url>::~StringHolder'. 
Strangely the value of 'str_' is definitely invalid. What is also counter-intuitive (for me) that the 'StringHolder(const std::initializer_list<std::string> args))' constructor is being used. @COM8 do you agree that this is the wrong constructor for this example?

**Comment by COM8** [comment]
For me this make sense since there is no other valid ctr that takes only a `const char*`, so I guess `StringHolder(const std::initializer_list<std::string> args)` is being chosen.

**Comment by adventurist** [comment]
Confirming I've also hit this

**Comment by ollien** [comment]
@COM8 yep, I'll take a look as soon as I can. I'm confused how this happened

**Comment by ollien** [comment]
Ok, I've done some digging, but I'm unsure about a full resolution at the moment (cpr was the first and only work I did with Conan). This warning is pretty prominently displayed when you go to install `cpr` with Conan.

```
************************* WARNING: GCC OLD ABI COMPATIBILITY ***********************
 
Conan detected a GCC version > 5 but has adjusted the 'compiler.libcxx' setting to
'libstdc++' for backwards compatibility.
Your compiler is likely using the new CXX11 ABI by default (libstdc++11).

If you want Conan to use the new ABI for the default profile, run:

    $ conan profile update settings.compiler.libcxx=libstdc++11 default

Or edit '/root/.conan/profiles/default' and set compiler.libcxx=libstdc++11

************************************************************************************
```

Changing the ABI as it suggests seems to make everything work as expected. It even makes some level of sense if you look at the [valgrind errors](https://gist.github.com/ollien/099f7176d91f5137025abf81b9325cb9), given what it's failing to read.

That said, the tests that run in the CCI build system seem to pass just fine with `libstdc++`, [if we look at some of the past runs](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/7367/1/summary.json)

Short answer: there is a workaround if this ABI change is acceptable to you. 

I don't have a longer answer. I'm going to file an issue in CCI about it.

**Comment by COM8** [comment]
Aha! Thaks for investigating @ollien

**Comment by HoMiGT** [comment]
Thank you very much. I thought there was something wrong with my code. It was a compilation problem @ollien
